### PR TITLE
feat: add rekey KMIP operation

### PR DESCRIPTION
**Pull request by Manuthor**
- add `ReKey` KMIP operation (to refresh symmetric key)
  * add impl for server and client `ckms`

### CODE REVIEW

File: crate/cli/src/actions/symmetric/keys/rekey.rs [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+            ..ReKey::default()
+        };
+        let unique_identifier = kms_rest_client
+            .rekey(rekey_key_request)
+            .await
+            .with_context(|| "failed rekeying the key")?
+            .unique_identifier
+            .to_string();
+
+        let mut stdout = console::Stdout::new("The symmetric key was successfully generated.");
```
Comment (ccorsin) [inline crate/cli/src/actions/symmetric/keys/rekey.rs:35]: `The symmetric key was successfully refreshed ` instead ?

File: crate/cli/src/tests/symmetric/rekey.rs [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+    // Compare the symmetric key bytes
+    let old_key = read_object_from_json_ttlv_file(&tmp_path.join("aes_sym"))?
+        .key_block()?
+        .key_bytes()?;
+    let new_key = read_object_from_json_ttlv_file(&tmp_path.join("aes_sym_2"))?
+        .key_block()?
+        .key_bytes()?;
+
+    assert_ne!(old_key, new_key);
```
Comment (ccorsin) [inline crate/cli/src/tests/symmetric/rekey.rs:93]: if attributes should be preserved on refresh - maybe create a symkey with attributes, and also check that the rekeyed symkey still have them ?

File: crate/server/src/core/operations/rekey.rs [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    owner: &str,
+    params: Option<&ExtraDatabaseParams>,
+) -> KResult<ReKeyResponse> {
+    trace!("ReKey: {}", serde_json::to_string(&request)?);
+
+    if request.protection_storage_masks.is_some() {
+        kms_bail!(KmsError::UnsupportedPlaceholder)
+    }
+
+    // there must be an identifier
```
Comment (ccorsin) [inline crate/server/src/core/operations/rekey.rs:31]: so it's not optionnal - cf. clap argument structure ?

File: crate/kmip/src/kmip/kmip_operations.rs [hunk 4]
```diff
@@ -1237,6 +1242,43 @@ pub struct RevokeResponse {
...
+///
+/// For the existing key, the server SHALL create a Link attribute of Link Type Replacement Object pointing to the replacement key. For the replacement key, the server SHALL create a Link attribute of Link Type Replaced Key pointing to the existing key.
+///
+/// An Offset MAY be used to indicate the difference between the Initial Date and the Activation Date of the replacement key. If no Offset is specified, the Activation Date, Process Start Date, Protect Stop Date and Deactivation Date values are copied from the existing key.
+#[derive(Debug, Serialize, Deserialize, Default, PartialEq, Eq)]
+#[serde(rename_all = "PascalCase")]
+pub struct ReKey {
+    // Determines the existing Symmetric Key being re-keyed. If omitted, then the ID Placeholder value is used by the server as the Unique Identifier.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub unique_identifier: Option<UniqueIdentifier>,
```
Comment (ccorsin) [inline crate/kmip/src/kmip/kmip_operations.rs:1259]: Remove option here?

File: crate/cli/src/actions/symmetric/keys/rekey.rs (hunk 1 above)
Comment (Manuthor) [reply crate/cli/src/actions/symmetric/keys/rekey.rs]: done in a38519480e509b77275d3d6cbb1e0017e3d4cc93

File: crate/cli/src/tests/symmetric/rekey.rs (hunk 2 above)
Comment (Manuthor) [reply crate/cli/src/tests/symmetric/rekey.rs]: done in a38519480e509b77275d3d6cbb1e0017e3d4cc93

File: crate/kmip/src/kmip/kmip_operations.rs (hunk 4 above)
Comment (Manuthor) [reply crate/kmip/src/kmip/kmip_operations.rs:1259]: I'd rather let it be optional be conform the KMIP specs

File: crate/server/src/core/operations/rekey.rs (hunk 3 above)
Comment (Manuthor) [reply crate/server/src/core/operations/rekey.rs:31]: well It is optionnal in KMIP https://docs.oasis-open.org/kmip/kmip-spec/v2.0/os/kmip-spec-v2.0-os.html#_Toc6497568

**Review by ccorsin — APPROVED** [review]

**Pull request merged**